### PR TITLE
Eliminate SDK client generics

### PR DESCRIPTION
**Pull request by jdisanti**
## Motivation and Context
The motivation here is to simplify implementation of waiters by reducing the amount of generic bounds required to implement them. The connector and middleware can still be customized since they are dynamic.

## Checklist
- [x] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/PaginatorGenerator.kt [hunk 1]
```diff
@@ -125,7 +129,7 @@ class PaginatorGenerator private constructor(
                 builder: #{Builder}
             }
 
-            impl${generics.inst} ${paginatorName}${generics.inst} where #{bounds:W} {
+            impl${generics.inst} ${paginatorName}${generics.inst} #{bounds:W} {
```
Comment (rcoh) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/PaginatorGenerator.kt:132]: nit: small inconsistency that one includes the `where` and the other does not

File: aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/AwsFluentClientDecorator.kt [hunk 2]
```diff
@@ -23,24 +24,52 @@ import software.amazon.smithy.rust.codegen.smithy.RuntimeConfig
...
+    }
+
+    /** Instantiation */
+    override val inst = ""
+
+    /** Trait bounds */
+    override val bounds = writable { }
+
+    /** Bounds for generated `send()` functions */
+    override fun sendBounds(input: Symbol, output: Symbol, error: RuntimeType): Writable = writable { }
```
Comment (rcoh) [inline aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/AwsFluentClientDecorator.kt:72]: I'm really surprised this compiles! It generates `fn send(...) where { body }`. Probably worth including the `where` in the bounds just so we generate a slightly more normal looking code

File: aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/AwsFluentClientDecorator.kt [hunk 3]
```diff
@@ -88,33 +112,34 @@ class AwsFluentClientDecorator : RustCodegenDecorator {
...
+                    E: Into<#{ConnectorError}>,
+                {
                     let retry_config = conf.retry_config.as_ref().cloned().unwrap_or_default();
                     let timeout_config = conf.timeout_config.as_ref().cloned().unwrap_or_default();
                     let sleep_impl = conf.sleep_impl.clone();
                     let mut builder = #{aws_smithy_client}::Builder::new()
-                        .connector(conn)
-                        .middleware(#{Middleware}::new());
+                        .connector(#{DynConnector}::new(conn))
+                        .middleware(#{DynMiddleware}::new(#{Middleware}::new()));
```
Comment (rcoh) [inline aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/AwsFluentClientDecorator.kt:142]: we _could_ consider sticking these types into the client generics for a little `DRY` but not sure if it's worth it

File: CHANGELOG.next.toml [hunk 4]
```diff
@@ -78,3 +78,12 @@ message = "MSRV increased from `1.54` to `1.56.1` per our 2-behind MSRV policy."
 references = ["smithy-rs#1130"]
 meta = { "breaking" = true, "tada" = false, "bug" = false }
 author = "jdisanti"
+
+[[aws-sdk-rust]]
+message = """
+`Client` no longer has generics, and now uses `DynConnector` and `DynMiddleware` to allow for connector/middleware customization.
```
Comment (rcoh) [inline CHANGELOG.next.toml:84]: can you update the changelog message to clarify that this is all generated service clients? Might also be worth calling out that this simplifies the code and docs. 

The only other call out might be custom retry policies? Not sure.

**Review by rcoh — APPROVED** [review]
LGTM! This is huge IMO.

File: aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/AwsFluentClientDecorator.kt (hunk 2 above)
Comment (jdisanti) [reply aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/AwsFluentClientDecorator.kt:72]: Wow, good catch! I'm also surprised 😅

**Comment by rcoh** [comment]
LGTM!

File: CHANGELOG.next.toml [hunk 5]
```diff
@@ -83,4 +83,61 @@ author = "jdisanti"
...
+    // ...
+}
+```
+
+Then the generics and trait bounds will no longer be necessary:
+```rust
+fn some_function(conn: DynConnector) -> Result<()> {
+    // ...
+}
+```
```
Comment (Velfi) [inline CHANGELOG.next.toml:136]: > I felt a great disturbance in the Force, as if millions of ~voices~ **`where` clauses** suddenly cried out in terror and were suddenly silenced. I fear something ~terrible~ **awesome** has happened.

**Review by Velfi — APPROVED** [review]

**Pull request merged**